packages/react/src/popover/viewport/PopoverViewport.tsx

```diff
@@ -0,0 +1,302 @@
+  useIsoLayoutEffect(() => {
+    // When a transition is in progress, we store the next content in capturedNodeRef.
+    // This handles the case where the trigger changes multiple times before the transition finishes.
+    // We want to always capture the latest content for the previous snapshot.
+    // So clicking quickly on T1, T2, T3 will result in the following sequence:
+    // 1. T1 -> T2: previousContent = T1, currentContent = T2
+    // 2. T2 -> T3: previousContent = T2, currentContent = T3
+    const source = currentContainerRef.current ?? nextContainerRef.current;
+    if (!source) {
+      return;
+    }
+
+    const wrapper = document.createElement('div');
+    for (const child of Array.from(source.childNodes)) {
+      wrapper.appendChild(child.cloneNode(true));
+    }
+
+    capturedNodeRef.current = wrapper;
+  });
+
+  const handleMeasureLayout = useEventCallback(() => {
+    currentContainerRef.current?.style.setProperty('animation', 'none');
+    currentContainerRef.current?.style.setProperty('transition', 'none');
+
+    previousContainerRef.current?.style.setProperty('display', 'none');
+
+    nextContainerRef.current?.style.setProperty('animation', 'none');
+    nextContainerRef.current?.style.setProperty('transition', 'none');
+  });
+
+  type MeasureLayoutCompleteData = {
+    previousDimensions: Dimensions | null;
+    nextDimensions: Dimensions;
+  };
+
+  const handleMeasureLayoutComplete = useEventCallback((data: MeasureLayoutCompleteData) => {
+    currentContainerRef.current?.style.removeProperty('animation');
+    currentContainerRef.current?.style.removeProperty('transition');
+
+    previousContainerRef.current?.style.removeProperty('display');
+
+    nextContainerRef.current?.style.removeProperty('animation');
+    nextContainerRef.current?.style.removeProperty('transition');
+
+    setPreviousContentDimensions(data.previousDimensions);
+  });
+
+  React.useEffect(() => {
+    floatingContext.events.on('measure-layout', handleMeasureLayout);
+    floatingContext.events.on('measure-layout-complete', handleMeasureLayoutComplete);
+
+    return () => {
+      floatingContext.events.off('measure-layout', handleMeasureLayout);
+      floatingContext.events.off('measure-layout-complete', handleMeasureLayoutComplete);
+    };
+  }, [floatingContext, handleMeasureLayout, handleMeasureLayoutComplete]);
+
+  const lastHandledTriggerRef = React.useRef<HTMLElement | null>(null);
+
+  useIsoLayoutEffect(() => {
+    // When a trigger changes, set the captured children HTML to state,
+    // so we can render both new and old content.
+    if (
+      open &&
+      activeTrigger &&
+      previousActiveTrigger &&
+      activeTrigger !== previousActiveTrigger &&
+      lastHandledTriggerRef.current !== activeTrigger &&
+      capturedNodeRef.current &&
+      previousContentNode?.innerHTML !== capturedNodeRef.current.innerHTML
```

Comment:
Even just reading `innerHTML` can prevent browser extensions from accepting code with it  iirc